File: app/src/main/java/com/mg/zeearchiver/dialogs/ExtractProgressDialog.java
package com.mg.zeearchiver.dialogs;

import android.app.AlertDialog;
import android.arch.core.util.Function;
import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import com.mg.zeearchiver.R;
import java.util.concurrent.Callable;

public class ExtractProgressDialog {
    private TextView currItem, percentage;
    private View root;
    final private Context con;
    private AlertDialog pd;
    final private Button cancelBtn;

    public ExtractProgressDialog(Context context, final Callable cancelAction) {
        con = context;
        root = LayoutInflater.from(context).inflate(R.layout.progress_dialog, null);
        currItem = root.findViewById(R.id.current_file);
        currItem.setEllipsize(TextUtils.TruncateAt.MARQUEE);
        currItem.setSelected(true);
        percentage = root.findViewById(R.id.comp_ratio);
        cancelBtn = root.findViewById(R.id.cancel_extraction_btn);
        cancelBtn.setVisibility(View.VISIBLE);
        cancelBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                cancelBtn.setEnabled(false);
                try {
                    cancelAction.call();
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
    }

    public View getRoot() {
        return root;
    }

    public void showDialog(String title) {
        AlertDialog.Builder builder = new AlertDialog.Builder(con);
        pd = builder.setView(root).setTitle(title)
                .setCancelable(false).create();
        pd.getWindow().setWindowAnimations(R.style.moving_dialog);
        pd.show();
    }

    public void dismiss() {
        if (pd != null && pd.isShowing())
            pd.dismiss();
    }

    public void setDialogTitle(String title) {
        pd.setTitle(title);
    }

    public void setDialogTitle(int resid) {
        pd.setTitle(resid);
    }

    public void setCurrentItemText(String st) {
        currItem.setText(st);
    }

    public void setPercentage(String st) {
        percentage.setText(st);
    }

    public void setPercentage(long percent) {
        //"Compression Ratio:"+ratio+"%"
        percentage.setText(con.getString(R.string.compression_ratio) + ":" + percent + "%");
    }
}

File: app/src/main/java/com/mg/zeearchiver/dialogs/ExtractProgressDialog.kt
package com.mg.zeearchiver.dialogs

import android.app.AlertDialog
import android.content.Context
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.widget.Button
import android.widget.TextView
import com.mg.zeearchiver.R
import java.util.concurrent.Callable

class ExtractProgressDialog(private val con: Context, cancelAction: Callable<*>) {
    private val currItem: TextView
    private val percentage: TextView
    val root: View
    private lateinit var pd: AlertDialog
    private val cancelBtn: Button
    fun showDialog(title: String?) {
        val builder = AlertDialog.Builder(con)
        pd = builder.setView(root).setTitle(title)
                .setCancelable(false).create()
        pd.window?.setWindowAnimations(R.style.moving_dialog)
        pd.show()
    }

    fun dismiss() {
        if (pd.isShowing) pd.dismiss()
    }

    fun setDialogTitle(title: String?) {
        pd.setTitle(title)
    }

    fun setDialogTitle(resid: Int) {
        pd.setTitle(resid)
    }

    fun setCurrentItemText(st: String?) {
        currItem.text = st
    }

    fun setPercentage(st: String?) {
        percentage.text = st
    }

    fun setPercentage(percent: Long) {
        //"Compression Ratio:"+ratio+"%"
        percentage.text = "${con.getString(R.string.compression_ratio)}: $percent %"
    }

    init {
        root = LayoutInflater.from(con).inflate(R.layout.progress_dialog, null)
        currItem = root.findViewById(R.id.current_file)
        currItem.ellipsize = TextUtils.TruncateAt.MARQUEE
        currItem.isSelected = true
        percentage = root.findViewById(R.id.comp_ratio)
        cancelBtn = root.findViewById(R.id.cancel_extraction_btn)
        cancelBtn.visibility = View.VISIBLE
        cancelBtn.setOnClickListener {
            it.isEnabled = false
            try {
                cancelAction.call()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}